processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,323 @@
+   *
+   * See StringDimensionIndexer.makeColumnValueSelector() for a reference implementation.
+   *
+   * @param spec Specifies the output name of a dimension and any extraction functions to be applied.
+   * @param currEntry Provides access to the current TimeAndDims object in the Cursor
+   * @param desc Descriptor object for this dimension within an IncrementalIndex
+   * @return A new object that reads rows from currEntry
+   */
+  public Object makeColumnValueSelector(
+      DimensionSpec spec,
+      IncrementalIndexStorageAdapter.EntryHolder currEntry,
+      IncrementalIndex.DimensionDesc desc
+  );
+
+
+  /**
+   * Compares the row values for this DimensionIndexer's dimension from a TimeAndDims key.
+   *
+   * The dimension values within a TimeAndDims key always use the "unsorted" ordering for encoded values.
+   *
+   * The row values are passed to this function as an Object, the implementer should cast them to the type
+   * appropriate for this dimension.
+   *
+   * For example, a dictionary encoded String implementation would cast the Objects as int[] arrays.
+   *
+   * When comparing, if the two arrays have different lengths, the shorter array should be ordered first.
+   *
+   * Otherwise, the implementer of this function should iterate through the unsorted encoded values, converting
+   * them to their actual type (e.g., performing a dictionary lookup for a dict-encoded String dimension),
+   * and comparing the actual values until a difference is found.
+   *
+   * Refer to StringDimensionIndexer.compareTimeAndDimsKey() for a reference implementation.
+   *
+   * @param lhs dimension value array from a TimeAndDims key
+   * @param rhs dimension value array from a TimeAndDims key
+   * @return comparison of the two arrays
+   */
+  public int compareTimeAndDimsKey(Object lhs, Object rhs);
+
+
+  /**
+   * Check if two row value arrays from TimeAndDims keys are equal.
+   *
+   * @param lhs dimension value array from a TimeAndDims key
+   * @param rhs dimension value array from a TimeAndDims key
+   * @return true if the two arrays are equal
+   */
+  public boolean checkTimeAndDimsKeyEqual(Object lhs, Object rhs);
+
+
+  /**
+   * Given a row value array from a TimeAndDims key, generate a hashcode.
+   * @param key dimension value array from a TimeAndDims key
+   * @return hashcode of the array
+   */
+  public int getTimeAndDimsKeyHashcode(Object key);
+
+
+  /**
+   * Given a row value array from a TimeAndDims key, as described in the documentatiion for compareTimeAndDimsKey(),
+   * convert the unsorted encoded values to a list or array of actual values.
+   *
+   * @param key dimension value array from a TimeAndDims key
+   * @param asList if true, return an array; if false, return a list
+   * @return list or array containing the actual values corresponding to the encoded values in the input array
+   */
+  public Object convertTimeAndDimsKeyToActualArrayOrList(Object key, boolean asList);
```

Comment:
That function actually can also return single values instead of array/list; I decided to keep it as is for now and update the documentation to indicate that it can return single values.

The reasoning being:
- The caller would have to do `instanceof int[]`, `instanceof float[]`, or use the methods from `java.lang.reflect.Array` which are slow, to determine whether to turn the array into a list if needed
- If this method always returned an array (no single values either), the caller would face the same type explosion when trying to retrieve values or have to use `java.lang.reflect.Array`